src/storage.rs

```diff
@@ -0,0 +1,175 @@
+use anyhow::{Error, bail};
+use aws_sdk_s3::error::SdkError;
+use aws_sdk_s3::operation::get_object::GetObjectError;
+use aws_sdk_s3::operation::head_object::HeadObjectError;
+use aws_sdk_s3::primitives::ByteStream;
+use reqwest::StatusCode;
+use std::path::Path;
+use tokio::runtime::Runtime;
+
+pub(crate) enum Storage {
+    ReadOnly(CdnReader),
+    ReadWrite(S3Storage),
+}
+
+impl Storage {
+    pub(crate) fn file_status(&self, path: &str) -> Result<FileStatus, Error> {
+        if let Some(hash) = self.get_file(&format!("{path}.sha256"))? {
+            Ok(FileStatus::Present {
+                sha256: hash.trim().to_string(),
+            })
+        } else if self.file_exists(path)? {
+            Ok(FileStatus::Legacy)
+        } else {
+            Ok(FileStatus::Missing)
+        }
+    }
+
+    pub(crate) fn upload_file(&self, path: &str, file: &Path) -> Result<(), Error> {
+        match self {
+            Storage::ReadOnly(_) => panic!("unsupported in read-only mode"),
+            Storage::ReadWrite(storage) => {
+                storage.runtime.block_on(
+                    storage
+                        .s3
+                        .put_object()
+                        .bucket(&storage.s3_bucket)
+                        .key(path)
+                        .body(storage.runtime.block_on(ByteStream::from_path(file))?)
+                        .send(),
+                )?;
+                Ok(())
+            }
+        }
+    }
+
+    pub(crate) fn write_contents(&self, path: &str, content: &[u8]) -> Result<(), Error> {
+        match self {
+            Storage::ReadOnly(_) => panic!("unsupported in read-only mode"),
+            Storage::ReadWrite(storage) => {
+                storage.runtime.block_on(
+                    storage
+                        .s3
+                        .put_object()
```

Comment:
I _think_ if we switch to https://docs.rs/aws-sdk-s3-transfer-manager/latest/aws_sdk_s3_transfer_manager/client/struct.Client.html#method.upload it'll auto-manage the uploads, including splitting into chunks, etc. (it's in preview but seems likely to be fine for our purposes, and probably will become stable soon)